app/code/Magento/GiftMessage/Test/Unit/Observer/SalesEventOrderItemToQuoteItemObserverTest.php

```diff
@@ -0,0 +1,225 @@
+     */
+    private $eventMock;
+
+    /**
+     * @var Order|MockObject
+     */
+    private $orderMock;
+
+    /**
+     * @var OrderItem|MockObject
+     */
+    private $orderItemMock;
+
+    /**
+     * @var Store|MockObject
+     */
+    private $storeMock;
+
+    /**
+     * @var MessageInterface|MockObject
+     */
+    private $messageMock;
+
+    /**
+     * @var QuoteItem|MockObject
+     */
+    private $quoteItemMock;
+
+    /**
+     * Prepare environment for test
+     */
+    public function setUp(): void
+    {
+        $this->messageFactoryMock = $this->getMockBuilder(MessageFactory::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['create'])
+            ->getMock();
+        $this->giftMessageHelperMock = $this->createMock(MessageHelper::class);
+        $this->observerMock = $this->createMock(Observer::class);
+        $this->eventMock = $this->createPartialMock(Event::class, ['getOrderItem', 'getQuoteItem']);
+        $this->orderItemMock = $this->createPartialMock(
+            OrderItem::class,
+            ['getOrder', 'getStoreId', 'getGiftMessageId']
+        );
+        $this->quoteItemMock = $this->createPartialMock(QuoteItem::class, ['setGiftMessageId']);
+        $this->orderMock = $this->createPartialMock(Order::class, ['getReordered']);
+        $this->storeMock = $this->createMock(Store::class);
+        $this->messageMock = $this->createMock(MessageModel::class);
+
+        $this->eventMock->expects($this->atLeastOnce())
+            ->method('getOrderItem')
+            ->willReturn($this->orderItemMock);
+
+        $this->orderItemMock->expects($this->atLeastOnce())
+            ->method('getOrder')
+            ->willReturn($this->orderMock);
+
+        $this->observerMock->expects($this->atLeastOnce())
+            ->method('getEvent')
+            ->willReturn($this->eventMock);
+
+        $objectManager = new ObjectManager($this);
+
+        $this->observer = $objectManager->getObject(
+            SalesEventOrderItemToQuoteItemObserver::class,
+            [
+                'messageFactory' => $this->messageFactoryMock,
+                'giftMessageMessage' => $this->giftMessageHelperMock
+            ]
+        );
+    }
+
+    /**
+     * Test when the order is reorder
+     */
+    public function testReorder()
+    {
+        $this->orderMock->expects($this->once())
+            ->method('getReordered')
+            ->willReturn(true);
+
+        $this->giftMessageHelperMock->expects($this->never())
+            ->method('isMessagesAllowed');
+
+        $this->eventMock
+            ->expects($this->never())
+            ->method('getQuoteItem')
+            ->willReturn($this->quoteItemMock);
+
+        /** Run observer */
+        $this->observer->execute($this->observerMock);
+    }
+
+    /**
+     * Test when the order is new reorder and gift message is not allowed
+     */
+    public function testNewOrderWhenGiftMessageIsNotAllowed()
```

Comment:
Also: `testNewOrderWhenGiftMessageIsNotAllowedDoesNotGetQuoteItem` to explain what actually is going on inside (It's very useful when you're browsing ouput of phpunit execution)